job_creator.php

```diff
@@ -89,7 +138,28 @@ private function getPhpVersion(int $phpIndex): string
             }
         }
         $phpVersions = INSTALLER_TO_PHP_VERSIONS[$key] ?? INSTALLER_TO_PHP_VERSIONS['4'];
-        return $phpVersions[$phpIndex] ?? $phpVersions[count($phpVersions) - 1];
+        // Use the max allowed php version
+        if (!array_key_exists($phpIndex, $phpVersions)) {
+            for ($i = count($phpVersions) - 1; $i >= 0; $i--) {
+                $phpVersion = $phpVersions[count($phpVersions) - 1];
+                if ($this->isAllowedPhpVersion($phpVersion)) {
+                    return $phpVersion;
+                }
+            }
```

Comment:
If we get to the end of this for loop and haven't returned, we should throw an error.
One will be thrown implicitly for us below but makes more sense to explicitly throw ourselves with a meaningful error message.

Reply:
The idea is that there's a couple of fallback methods to get a PHP version below, rather than throwing an error message early.  It's more important we find a usable php version for the matrix job, rather than throwing an exception that will fail the job if we cannot get the exact php version we specified.

Reply:
Hmmm what will `array_splice($phpVersions, $phpIndex)` do if the index doesn't exist? I instinctively assumed it would throw an error but I now think that was an incorrect assumption

Reply:
ah it's a typo, it's meant to be array_slice in both instances

Reply:
have updated

Reply:
If the array key doesn't exist, the two array_slice checks are guaranteed to fail if we've gone through all available PHP versions and they aren't allowed - but there's no need to optimise this, it's okay as is.